Add new `Rails/ForeignKeyName` cop

fatkodima opened the pull request:
https://rails.rubystyle.guide/#meaningful-foreign-key-naming

pirj commented on config/default.yml:
```diff
@@ -272,6 +272,14 @@ Rails/FindEach:
+  Enabled: pending
+  VersionAdded: '2.8'
+  Include:
+    - db/migrate/*.rb
```
This will only work with migrations of the primary database.

pirj commented on lib/rubocop/cop/rails/foreign_key_name.rb:
```diff
@@ -0,0 +1,43 @@
+          foreign_key_name_candidate?(node) do |options_node|
+            options_node = options_node.first
+
+            add_offense(node) if options_node.nil? || !includes_name?(options_node)
```
Is it hard to add auto-correction? Will you have to catch the `on_block` for `t.foreign_key` to get the table name from `create_table`/`change_table`?

pirj commented on lib/rubocop/cop/rails/foreign_key_name.rb:
```diff
@@ -0,0 +1,43 @@
+        private
+
+        def includes_name?(node)
+          node.pairs.any? { |p| p.key.value.to_sym == :name }
```
I think there's now an option in `rubocop-ast` to match any pair key https://docs.rubocop.org/rubocop-ast/node_pattern.html#for-match-in-any-order
Example https://github.com/rubocop/rubocop-rspec/blob/a8101aeb8607d9265414502aed16b459a077feef/lib/rubocop/cop/rspec/multiple_expectations.rb#L60

pirj reviewed (commented):
Can you please rebase to re-trigger CI?

fatkodima commented:
It's been a while 😄 

@pirj, thank you for suggestions! I addressed all of them, please take a look.

pirj commented on config/default.yml:
```diff
@@ -368,6 +368,14 @@ Rails/FindEach:
+  Description: 'Name your foreign keys explicitly instead of relying on Rails auto-generated FK names.'
+  StyleGuide: 'https://rails.rubystyle.guide/#meaningful-foreign-key-naming'
+  Enabled: pending
+  VersionAdded: '2.8'
```
2.13 ?

pirj commented on lib/rubocop/cop/rails/foreign_key_name.rb:
```diff
@@ -0,0 +1,73 @@
+          end
+
+          from_table = from_table(node)
+          "#{from_table}_#{column}_fk"
```
What if we return a Symbol here, and then just `insert_after(node, ", name: foreign_key_name)` above with no string interpolation and `:"` conversion?

fatkodima replied:
Tried your suggestion:
```ruby
insert_after(node, name: foreign_key_name)
```

But got this:
```ruby
+add_foreign_key :articles, :authors{:name=>:articles_author_id_fk}
```

It just appended "inspected" value. Seems like we need to explicitly build an appended source string.

pirj commented:
@fatkodima Do you think it makes sense to run the new cop against https://github.com/eliotsykes/real-world-rails to check if there are errors, and check for false positives?
I can do this for you, as it's a tedious task.

@koic Will you please take a look?

pirj reviewed (commented):
One concern I have about this cop is that once introduced, it will start flagging old migrations.
And fixing old migrations means renaming FKs in the new migrations. That wouldn't appease the cop, will it?
Is there a way to set some date so that the cop is not triggered by old migrations?

fatkodima commented:
> Do you think it makes sense to run the new cop against https://github.com/eliotsykes/real-world-rails to check if there are errors, and check for false positives?

👍  I ran it on gitlab's codebase. Got 1 error, fixed it. For other migration files - seems like is working as expected.

> One concern I have about this cop is that once introduced, it will start flagging old migrations.

I added a new option to be able to ignore old migrations. I think, it will be also useful for other existing cops, like `AddColumnIndex` or `CreateTableWithTimestamps`, etc.

pirj commented on CHANGELOG.md:
```diff
@@ -2,6 +2,10 @@
 
+### New features
+
+* [#277](https://github.com/rubocop/rubocop-rails/pull/277): Add new `Rails/ForeignKeyName` cop. ([@fatkodima][])
```
Would you please change to the new changelog format? https://github.com/rubocop/rubocop-rails/commit/3cae05d8cf1a94e76496a42bf38e033f941690f6#diff-a66abd3d7020d7e9f081637dbc85308d8cd44db8e7eec508b9314f3664ff6968

pirj commented on config/default.yml:
```diff
@@ -368,6 +368,15 @@ Rails/FindEach:
+  Description: 'Name your foreign keys explicitly instead of relying on Rails auto-generated FK names.'
+  StyleGuide: 'https://rails.rubystyle.guide/#meaningful-foreign-key-naming'
+  Enabled: pending
+  VersionAdded: '2.13'
```
`'<<next>>'`

pirj commented on docs/modules/ROOT/pages/cops.adoc:
```diff
@@ -52,6 +52,7 @@ based on the https://rails.rubystyle.guide/[Rails Style Guide].
 * xref:cops_rails.adoc#railsfindby[Rails/FindBy]
 * xref:cops_rails.adoc#railsfindbyid[Rails/FindById]
 * xref:cops_rails.adoc#railsfindeach[Rails/FindEach]
+* xref:cops_rails.adoc#railsforeignkeyname[Rails/ForeignKeyName]
```
Please revert doc changes - they will be auto-generated on release.

pirj commented on spec/rubocop/cop/rails/foreign_key_name_spec.rb:
```diff
@@ -0,0 +1,102 @@
+
+  context 'StartAfterMigrationVersion is used' do
+    let(:cop_config) do
+      { 'StartAfterMigrationVersion' => 20_211_007_000_001 }
```
What if you define it as a string here? `"2021_10_07_00_00_01" is a tad easier to read 😄

pirj reviewed (approved):
Looks good, thank you!

fatkodima commented:
Updated.

pirj commented on docs/modules/ROOT/pages/cops_rails.adoc:
```diff
@@ -1893,6 +1893,64 @@ User.order(:foo).each
 
 * https://rails.rubystyle.guide#find-each
 
+== Rails/ForeignKeyName
```
Same here for docs - they will be automatically generated later.

fatkodima replied:
Nice, this has been changed since I actively contributed a year ago 👍
